File: python/3106. Lexicographically Smallest String After Operations With Constraint.py
class Solution:
    def getSmallestString(self, s: str, k: int) -> str:
        answer = ""
        for c in s:
            if k == 0:
                answer += c
                continue
            if k < ord('z')+1-ord(c):
                answer += chr(max(ord('a'), ord(c)-k))
                k = max(0, k-ord(c)+ord('a'))
            else:
                answer += 'a'
                if ord(c)-ord('a') < ord('z')+1-ord(c) and k >= ord(c)-ord('a'):
                    k -= ord(c)-ord('a')
                else:
                    k -= ord('z')+1-ord(c)
        return answer

File: java/3106. Lexicographically Smallest String After Operations With Constraint.java
class Solution {
    public String getSmallestString(String s, int k) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (k == 0) {
                sb.append(c);
                continue;
            }
            if (k < 'z'+1-c) {
                if (k >= c-'a') {
                    sb.append('a');
                    k -= c-'a';
                } else {
                    sb.append((char)(c-k));
                    k = 0;
                }
                continue;
            }
            sb.append('a');
            if (c-'a'<'z'+1-c && k >= c-'a') {
                k -= c-'a';
            } else {
                k -= 'z'+1-c;
            }
        }
        return sb.toString();
    }
}